prow/pluginhelp/hook/hook.go

```diff
@@ -0,0 +1,366 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package hook provides the plugin help components to be compiled into the hook binary.
+// This includes the code to fetch help from normal and external plugins and the code to build and
+// serve a pluginhelp.Help struct.
+package hook
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/kubernetes/pkg/util/sets"
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/pluginhelp"
+	"k8s.io/test-infra/prow/pluginhelp/externalplugins"
+	"k8s.io/test-infra/prow/plugins"
+)
+
+// TODO: unit test to ensure that external plugins with the same name have the same endpoint and events.
```

Comment:
I think this won't be needed if we invert the way we configure external plugins to something like:
```yaml
external_plugins:
  cherrypick:
    repos:
    - openshift/origin
    - openshift/origin-aggregated-logging
   events:
   - issue_comment
   - pull_request
```
Seems less verbose than what we have today:
```yaml
external_plugins:
  openshift/origin:
  - name: cherrypick
    events:
    - issue_comment
    - pull_request
  openshift/origin-aggregated-logging:
  - name: cherrypick
    events:
    - issue_comment
    - pull_request
```

Reply:
From our plugin config: https://github.com/openshift/release/blob/master/cluster/ci/config/prow/plugins.yaml

Reply:
I like that the suggested format makes it impossible to make two external plugins with the same name, but I don't like that it is opposite from how normal plugins are configured. Also once there are more external plugins this could make it difficult to determine what is enabled on each repo.

Another alternative we could consider is mapping repos to external plugin name (like how normal plugins are configured) and including a separate map from external plugin names to the external plugin endpoint and events. (This would need a trivial unit test to make sure that all external plugin names in the repo mapping are associated with external plugin endpoint/events in the second mapping.) This is less verbose than the current format, would still prevent duplicate external plugins, and would make it easy to tell which external plugins are enabled on a repo.

Reply:
If we default events, the current approach could become bearable.